File: packages/captions/src/layout.ts
/**
 * Caption placement relative to the VIDEO rect inside the output frame.
 *
 * The output frame (canvas) is the video plus padding plus any letterbox bars.
 * At Offset 0 a top/bottom caption sits at the video's edge (in the padding, not
 * covering the video); positive Offset moves it INWARD over the video and negative
 * tucks it OUTWARD into the padding. The baseline is clamped to the frame edge so a
 * full-bleed video keeps the whole Offset range live instead of dead-clamping.
 *
 * Both renderers derive from this: the CSS preview overlay and the Rust ASS
 * generator (which mirrors `captionHeightFrac` / `captionTopFrac`). All values
 * are fractions of the canvas (0..1). Keep the two in sync.
 */

export interface VideoRect {
	/** Video rect within the canvas, as fractions of canvas width/height. */
	left: number;
	right: number;
	top: number;
	bottom: number;
}

/** Largest caption block height we ever reserve, as a fraction of the frame.
 *  Keeps the clamp from pushing captions past the frame centre. */
const MAX_CAP_FRAC = 0.7;
/** Line-height + breathing room factor for the height estimate. */
const LINE_FACTOR = 1.35;

/** Estimated caption block height as a fraction of frame height. Uses `maxLines`
 *  as an upper bound so the clamp reserves enough room for the tallest case. */
export function captionHeightFrac(fontSizePct: number, maxLines: number): number {
	const lines = Math.max(1, maxLines);
	return Math.min(MAX_CAP_FRAC, (fontSizePct / 100) * lines * LINE_FACTOR);
}

/**
 * Fraction-from-top of the caption block's TOP edge (the block grows downward).
 * `null` means centre: vertically centred on the video (which is itself centred
 * in the canvas), handled by the caller. `capFrac` comes from
 * {@link captionHeightFrac}.
 */
export function captionTopFrac(
	position: "top" | "center" | "bottom",
	offsetPct: number,
	capFrac: number,
	video: Pick<VideoRect, "top" | "bottom">,
): number | null {
	if (position === "center") return null;
	// Signed against the on-frame edge, so every slider value is live; anchoring on the raw video edge dead-clamped the positive range at full bleed.
	const offset = offsetPct / 100;
	const cap = Math.max(0, Math.min(MAX_CAP_FRAC, capFrac));
	const maxTop = Math.max(0, 1 - cap);
	if (position === "bottom") {
		// Baseline just below the video (clamped on-frame); positive lifts up.
		const base = Math.min(video.bottom, maxTop);
		return Math.max(0, Math.min(base - offset, maxTop));
	}
	// Top: baseline just above the video (clamped on-frame); positive pushes down.
	const base = Math.max(0, video.top - cap);
	return Math.max(0, Math.min(base + offset, maxTop));
}

File: crates/recast-captions/src/layout.rs
use crate::model::{CaptionAnimation, CaptionStyle, TranscriptWord};

/// Largest caption block height ever reserved, as a fraction of the frame, so
/// the clamp cannot push a caption past the frame centre.
pub const MAX_CAP_FRAC: f64 = 0.7;

/// Line height plus breathing room, for the block-height estimate only.
const LINE_FACTOR: f64 = 1.35;

/// Estimated caption block height as a fraction of frame height. Uses
/// `max_lines` as an upper bound so the clamp reserves room for the tallest case.
pub fn caption_height_frac(font_size_pct: f64, max_lines: u32) -> f64 {
    (font_size_pct / 100.0 * max_lines.max(1) as f64 * LINE_FACTOR).min(MAX_CAP_FRAC)
}

/// Fraction-from-top of the caption block's top edge, the block growing
/// downward. `None` means centre, which the caller resolves against the video.
///
/// The baseline is the clamped ON-FRAME edge rather than the raw video edge:
/// anchoring on the raw edge left the whole positive offset range dead-clamped
/// whenever the video reached the frame edge.
pub fn caption_top_frac(
    position: &str,
    offset_pct: f64,
    cap: f64,
    v_top: f64,
    v_bottom: f64,
) -> Option<f64> {
    if position == "center" {
        return None;
    }
    // Signed: positive moves the caption inward over the video, negative tucks
    // it outward into the padding.
    let offset = offset_pct / 100.0;
    let cap = cap.clamp(0.0, MAX_CAP_FRAC);
    let max_top = (1.0 - cap).max(0.0);
    if position == "bottom" {
        let base = v_bottom.min(max_top);
        Some((base - offset).clamp(0.0, max_top))
    } else {
        let base = (v_top - cap).max(0.0);
        Some((base + offset).clamp(0.0, max_top))
    }
}

/// Groups `words` into the runs shown on screen at once.
pub fn chunk_words<'a>(
    words: &'a [TranscriptWord],
    anim: &CaptionAnimation,
) -> Vec<&'a [TranscriptWord]> {
    if words.is_empty() {
        return Vec::new();
    }
    let size = match anim.chunk.as_str() {
        "line" => words.len(),
        "word" => 1,
        _ => (anim.chunk_size as usize).max(1),
    };
    words.chunks(size).collect()
}

/// Greedy line break by character count, never splitting inside a word and
/// capped at `max_lines`. Returns groups of indices into `words`.
///
/// Measurement-free on purpose: the DOM and libass shape text differently, so a
/// break derived from either would drift. Both honour the break decided here.
pub fn break_into_lines(words: &[TranscriptWord], max_chars: u32, max_lines: u32) -> Vec<Vec<usize>> {
    let limit = max_chars.max(1) as usize;
    let cap = max_lines.max(1) as usize;
    let mut lines: Vec<Vec<usize>> = Vec::new();
    let mut current: Vec<usize> = Vec::new();
    let mut current_len = 0usize;
    for (i, w) in words.iter().enumerate() {
        let word_len = w.text.chars().count();
        let added = if current.is_empty() {
            word_len
        } else {
            current_len + 1 + word_len
        };
        if !current.is_empty() && added > limit {
            lines.push(std::mem::take(&mut current));
            current.push(i);
            current_len = word_len;
        } else {
            current.push(i);
            current_len = added;
        }
        if lines.len() == cap {
            break;
        }
    }
    if !current.is_empty() && lines.len() < cap {
        lines.push(current);
    }
    lines.truncate(cap);
    lines
}

/// Pill geometry from the style plus a measured text width. No measurement of
/// its own: the caller supplies `text_width_px`, so the DOM and the compositor
/// derive the same box from their own shapers.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PillBox {
    pub width: f64,
    pub height: f64,
    pub radius: f64,
    pub pad_x: f64,
    pub pad_y: f64,
}

pub fn pill_box(style: &CaptionStyle, font_px: f64, text_width_px: f64, line_count: usize) -> PillBox {
    let pad_x = style.box_padding_x_em * font_px;
    let pad_y = style.box_padding_y_em * font_px;
    let lines = line_count.max(1) as f64;
    let height = lines * style.line_height * font_px + 2.0 * pad_y;
    PillBox {
        width: text_width_px + 2.0 * pad_x,
        height,
        radius: (style.box_radius_em * font_px).min(height / 2.0).max(0.0),
        pad_x,
        pad_y,
    }
}
